File: ReactlikeMvvmVb/HiViewModel/HiViewModelBase.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel

Public Class HiViewModelBase
    Implements INotifyPropertyChanged, IHiObservador(Of String)
    Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

    Public Sub Atualizar(arg As String) Implements IHiObservador(Of String).Atualizar
        NotificarPropertyChanged(arg)
    End Sub

    Private Sub NotificarPropertyChanged(arg As String)
        If PropertyChangedEvent IsNot Nothing Then
            PropertyChangedEvent.Invoke(Me, New PropertyChangedEventArgs(arg))
        End If
    End Sub

    Public Function UseState(Of T)(valorInicial As T, nomePropd As String) As HiEstado(Of T)
        Dim hiEstado = New HiEstado(Of T)(valorInicial, nomePropd)
        hiEstado.Inscrever(Me)
        Return hiEstado
    End Function

    Public Function UseEffect(Of T)(fnCalcularValor As Func(Of T), deps As IEnumerable(Of HiSujeitoBase(Of String)), nomePropd As String) As HiEstadoDerivado(Of T)
        Dim hiEstadoDerivado = New HiEstadoDerivado(Of T)(fnCalcularValor, deps, nomePropd)
        hiEstadoDerivado.Inscrever(Me)
        Return hiEstadoDerivado
    End Function
End Class

File: ReactlikeMvvm/HiViewModel/HiViewModelBase.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using ReactlikeMvvm.HiPadraoObservador;

namespace ReactlikeMvvm.HiViewModel
{
    public class HiViewModelBase : INotifyPropertyChanged, IHiObservador<string>
    {
        public event PropertyChangedEventHandler? PropertyChanged;
        public void Atualizar(string arg)
        {
            NotificarPropertyChanged(arg);
        }
        private void NotificarPropertyChanged(string arg)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(arg));
        }
        public HiEstado<T> UseState<T>(T valorInicial, string nomePropd)
        {
            var hiEstado = new HiEstado<T>(valorInicial, nomePropd);
            hiEstado.Inscrever(this);
            return hiEstado;
        }
        public HiEstadoDerivado<T> UseEffect<T>(Func<T> fnCalcularValor, IEnumerable<HiSujeitoBase<string>> deps, string nomePropd)
        {
            var hiEstadoDerivado = new HiEstadoDerivado<T>(fnCalcularValor, deps, nomePropd);
            hiEstadoDerivado.Inscrever(this);
            return hiEstadoDerivado;
        }
    }
}